### PR TITLE
Change chess-python API to version 1.5

### PR DESCRIPTION
**Pull request by karol-brejna-i**
This PR upgrades python-chess package to the newest version and adds some basic regression tests.

### CODE REVIEW

**Comment by vitogit** [comment]
Pretty cool, awesome work @karol-brejna-i , I will try to review it this weekend as I want to take a better look.

**Comment by vitogit** [comment]
I run this with 3 games that in the previous version produced 1 or 2 tactics (I think is not deterministic) But this code produce 0 tactics
Here are the games
```
[Event "<1700 Rapid Arena"]
[Site "https://lichess.org/13Q4NqCj"]
[Date "2019.11.01"]
[White "PedrosF1"]
[Black "artemchampion"]
[Result "1-0"]
[UTCDate "2019.11.01"]
[UTCTime "08:47:32"]
[WhiteElo "1417"]
[BlackElo "1678"]
[WhiteRatingDiff "+12"]
[BlackRatingDiff "-11"]
[Variant "Standard"]
[TimeControl "600+0"]
[ECO "B40"]
[Termination "Normal"]

1. e4 c5 2. Nf3 e6 3. Nc3 a6 4. d4 b5 5. dxc5 Bxc5 6. Bd3 Qb6 7. O-O Bb7 8. Ne5 Nf6 9. a3 Nc6 10. Nxc6 Bxc6 11. b4 Bd4 12. Bd2 Nxe4 13. Bxe4 Bxe4 14. Nxe4 Bxa1 15. Qxa1 Rc8 16. Qxg7 Rf8 17. Nf6+ Ke7 18. Bg5 Kd6 19. Rd1+ Kc6 20. Nxd7 Rcd8 21. Bxd8 Rxd8 22. Qxf7 Qc7 23. Qxe6+ Kb7 24. Nc5+ Ka8 25. Rxd8+ Qxd8 26. Qc6+ Kb8 27. Qb7# 1-0

[Event "U1700 Rapid Arena"]
[Site "https://lichess.org/2BZx5TDD"]
[Date "2019.11.01"]
[Round "-"]
[White "Dicazman"]
[Black "Meju"]
[Result "0-1"]
[UTCDate "2019.11.01"]
[UTCTime "08:42:29"]
[WhiteElo "1547"]
[BlackElo "1629"]
[WhiteRatingDiff "-6"]
[BlackRatingDiff "+6"]
[Variant "Standard"]
[TimeControl "600+0"]
[ECO "C41"]
[Termination "Normal"]

1. e4 e5 2. Nf3 d6 3. d4 exd4 4. Nxd4 Nc6 5. Bb5 Bd7 6. Bxc6 Bxc6 7. Nc3 Nf6 8. Bg5 Be7 9. Bxf6 Bxf6 10. Nxc6 bxc6 11. O-O O-O 12. Re1 Be5 13. g3 Qd7 14. Rb1 Rfe8 15. Na4 Re6 16. Nc5 Qe7 17. Nxe6 Qxe6 18. b3 Qh3 19. Qf3 Re8 20. a4 c5 21. a5 Bc3 22. Re2 Bxa5 23. Ra1 Bb6 24. Qf5 Qh6 25. Qd7 Kf8 26. c4 Qh5 27. Re3 Qe5 28. Rd1 h5 29. f4 Qb2 30. f5 Qf6 31. e5 dxe5 32. Rxe5 Qxe5 33. Rf1 Qe3+ 34. Kh1 Qxb3 35. f6 g6 36. Qd5 c6 37. Qxc6 Qxc4 38. Qd6+ Kg8 39. Rd1 Qe4+ 40. Kg1 c4+ 41. Kf1 Qf3# 0-1

[Event "<1700 Rapid Arena"]
[Site "https://lichess.org/oFKr1UUx"]
[Date "2019.11.01"]
[White "fibberybosch"]
[Black "momentum"]
[Result "1-0"]
[UTCDate "2019.11.01"]
[UTCTime "08:41:29"]
[WhiteElo "1566"]
[BlackElo "1374"]
[WhiteRatingDiff "+4"]
[BlackRatingDiff "-3"]
[Variant "Standard"]
[TimeControl "600+0"]
[ECO "C21"]
[Termination "Normal"]

1. e4 e5 2. d4 exd4 3. c3 Nc6 4. cxd4 Bb4+ 5. Nc3 Bxc3+ 6. bxc3 d5 7. e5 Be6 8. Nf3 Nf6 9. exf6 gxf6 10. Bh6 Qe7 11. Be2 Bg4 12. O-O Bxf3 13. Bxf3 O-O-O 14. Rb1 Qd6 15. Qa4 a6 16. c4 dxc4 17. Qxc4 f5 18. d5 Ne5 19. Qf4 Nxf3+ 20. Qxf3 Qxh6 21. Rfc1 Rhg8 22. Qb3 b6 23. Qf3 Qg6 24. g3 h5 25. h4 Qg4 26. Qc3 Rd7 27. Rxb6 Rg6 28. Rb4 Qh3 29. Qh8+ Rd8 30. Rb8+ Kxb8 31. Qxd8+ Kb7 32. Rb1+ Rb6 33. Rc1 a5 34. Qxc7+ Ka6 35. Qc8+ Kb5 36. Rb1+ Ka4 37. Qc2+ Ka3 38. Qc3+ Ka4 39. Rxb6 Qg4 40. Qc2+ Ka3 41. Qb3# 1-0

```

Here are the tactics
```
[Event "<1700 Rapid Arena"]
[Site "https://lichess.org/13Q4NqCj"]
[Date "2019.11.01"]
[Round "?"]
[White "PedrosF1"]
[Black "artemchampion"]
[Result "1-0"]
[BlackElo "1678"]
[BlackRatingDiff "-11"]
[ECO "B40"]
[FEN "r3k2r/3p1ppp/pqb1p3/1p6/1P1bB3/P1N5/2PB1PPP/R2Q1RK1 b kq - 0 13"]
[SetUp "1"]
[Termination "Normal"]
[TimeControl "600+0"]
[UTCDate "2019.11.01"]
[UTCTime "08:47:32"]
[Variant "Standard"]
[WhiteElo "1417"]
[WhiteRatingDiff "+12"]

13... Bxe4 14. Nxe4 Bxa1 15. Qxa1 O-O 1-0

[Event "U1700 Rapid Arena"]
[Site "https://lichess.org/2BZx5TDD"]
[Date "2019.11.01"]
[Round "-"]
[White "Dicazman"]
[Black "Meju"]
[Result "0-1"]
[BlackElo "1629"]
[BlackRatingDiff "+6"]
[ECO "C41"]
[FEN "4r1k1/p1p2ppp/1b1p4/2p5/4P3/1P3QPq/2P1RP1P/R5K1 w - - 2 24"]
[SetUp "1"]
[Termination "Normal"]
[TimeControl "600+0"]
[UTCDate "2019.11.01"]
[UTCTime "08:42:29"]
[Variant "Standard"]
[WhiteElo "1547"]
[WhiteRatingDiff "-6"]

24. Qf5 Qxf5 25. exf5 Rxe2 26. c4 0-1

``

**Comment by karol-brejna-i** [comment]
I had similar experiences (the engine returned different scores between runs). I'll check the case.

**Comment by karol-brejna-i** [comment]
The engines produce a bit different scores so "investigate" doesn't count them as candidates for puzzles.
Maybe default engine settings in chess-python had changed... I'll check that.
I'll write some simple tests to check against the same scores, too.

**Comment by karol-brejna-i** [comment]
It is definitely sth with the new code. Will work on that and get back to you.

**Comment by karol-brejna-i** [comment]
I've created a branch (from master, https://github.com/vitogit/pgn-tactics-generator/compare/master...karol-brejna-i:producing-test-context?expand=1) to produce some test data on the older version.
I'll use them to validate the same functions in new versions.
Maybe when I finish the test we can decide if we want to include them (the tests) to the main codebase...

**Comment by vitogit** [comment]
> I've created a branch (from master, https://github.com/vitogit/pgn-tactics-generator/compare/master...karol-brejna-i:producing-test-context?expand=1) to produce some test data on the older version.
> I'll use them to validate the same functions in new versions.
> Maybe when I finish the test we can decide if we want to include them (the tests) to the main codebase...

Given the nondeterministic nature of engines not sure how easy/hard would be to investigate it.  One idea is to use single threading when running the engine  `--threads=1` to get more consistent results

**Comment by karol-brejna-i** [comment]
What I did was dump `puzzle` object (that contains analyzed positions including their scoring) to a json file.
I am using the file to recreate objects (without the need to use actual UCI engine) and then run some methods/functions, for example  `ambiguous()`, `is_complete()`. 

@vitogit could you take a look at the branch https://github.com/karol-brejna-i/pgn-tactics-generator/tree/producing-test-context and the following lines of ambiguous method: https://github.com/vitogit/pgn-tactics-generator/blob/master/modules/puzzle/position_list.py#L146-L147
```python
            if self.analysed_legals[0].evaluation.mate is None or self.analysed_legals[1].evaluation.cp is None:
                return
```
This method is supposed to return a boolean value. These lines return nothing.
(This is one of the reasons the results differ -- upgraded version vs current version -- but not the only one, unfortunately).

I assume that the intent was to return True here. Could you confirm?

**Comment by vitogit** [comment]
> What I did was dump `puzzle` object (that contains analyzed positions including their scoring) to a json file.
> I am using the file to recreate objects (without the need to use actual UCI engine) and then run some methods/functions, for example `ambiguous()`, `is_complete()`.
> 
> @vitogit could you take a look at the branch https://github.com/karol-brejna-i/pgn-tactics-generator/tree/producing-test-context and the following lines of ambiguous method: https://github.com/vitogit/pgn-tactics-generator/blob/master/modules/puzzle/position_list.py#L146-L147
> 
> ```python
>             if self.analysed_legals[0].evaluation.mate is None or self.analysed_legals[1].evaluation.cp is None:
>                 return
> ```
> 
> This method is supposed to return a boolean value. These lines return nothing.
> (This is one of the reasons the results differ -- upgraded version vs current version -- but not the only one, unfortunately).
> 
> I assume that the intent was to return True here. Could you confirm?

That line was added in your previous PR https://github.com/vitogit/pgn-tactics-generator/commit/23f543c6e5fc387ad4a385aa5060539d7f56e988#diff-e31f1011e582ea3d502823f13a4a64fdef87e41ed47178a0a6e497df1402a27cR146 
Probably that whole function can be refactored but looking at the previous code I think you want to return `false` because you inverted the `if` from 
```
self.analysed_legals[0].evaluation.mate is not None and self.analysed_legals[1].evaluation.cp is not None
```
to 
```
self.analysed_legals[0].evaluation.mate is None or self.analysed_legals[1].evaluation.cp is None:
```
or leave it as it was before

**Comment by karol-brejna-i** [comment]
Jeez, it looks like you are right. I broke this. Let me fix this first with a PR.

**Comment by karol-brejna-i** [comment]
I've created #24 to restore the condition.

Going back to test this evening.

**Comment by karol-brejna-i** [comment]
With ambiguous fixed, the code works as expected. Regression tests that I added confirm that `ambiguous()`, `is_complete()` and `investigate()` return the same results in both versions. The games are produced similarly.

File: modules/decoding.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+
+def score_from_dict(d) -> Score:
+    if not d:
+        return None
+    if d[0]:
+        return Cp(d[0])
+    else:
+        return Mate(d[1])
+    # return Score(d[0], d[1]) if d else None
```
Comment (vitogit) [inline modules/decoding.py:18]: Probably delete this comment

Comment (karol-brejna-i) [reply modules/decoding.py:18]: I’ll do that and rebase against the current master.

**Comment by karol-brejna-i** [comment]
I made some cleanups. Looks like this one is finally ready to be merged.
@vitogit , please, take a look!

File: modules/puzzle/position_list.py [hunk 2]
```diff
@@ -109,7 +111,7 @@ def is_complete(self, category, color, first_node, first_val):
                 if (self.material_difference() > 0.2
                         and abs(self.material_difference() - first_val) > 0.1
                         and first_val < 2
-                        and self.evaluation.mate is None
+                        and self.evaluation.mate() is None
```
Comment (vitogit) [inline modules/puzzle/position_list.py:114]: I see that in other place you used not is_mate() and here mate() is None, is there any difference? Probably best to always use the same to keep consistency. Not sure what do you think

File: positions_for_investigation.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python
+
+"""
+This utility allows for generating some data that can be usesd in tests.
```
Comment (vitogit) [inline positions_for_investigation.py:4]:
```suggestion
This utility allows for generating some data that can be used in tests.
```

File: test/unit/test_regression.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
+import json
```
Comment (vitogit) [inline test/unit/test_regression.py:1]: This is great. What do you think about adding a small section in the readme with some info about the tests, how to run them, etc

Comment (karol-brejna-i) [reply test/unit/test_regression.py:1]: I was thinking about a dedicated document about development aspect (how to build and publish a package -- in the future, how to test, etc.) This way we'd had a readme that is all about usage of the software, and other that is focused on development/testing. I'll start one today and share.

File: modules/puzzle/position_list.py (hunk 2 above)
Comment (karol-brejna-i) [reply modules/puzzle/position_list.py:114]: I intended for the code to look very similar to previous version to easily spot changes and do the refactoring in the next iteration. I would keep the code as it is for now. What do you think?

File: test/unit/test_regression.py (hunk 4 above)
Comment (karol-brejna-i) [reply test/unit/test_regression.py:1]: take a look at https://github.com/karol-brejna-i/pgn-tactics-generator/blob/chess-python-upgrade/DEVELOPMENT.md

Comment (vitogit) [reply test/unit/test_regression.py:1]: Excellent I like it.

File: modules/puzzle/position_list.py (hunk 2 above)
Comment (vitogit) [reply modules/puzzle/position_list.py:114]: Sounds good.

File: DEVELOPMENT.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+## Running the tests
+The test can be run with:
+```bash
+python -m unittest
+```
+
+The recommended way to run the test is using `pytest`:
+```bash
+pytest
```
Comment (vitogit) [inline DEVELOPMENT.md:30]: This is not in the requirements. Maybe add a line about how to install it
```
pip3 install -U pytest
```

File: DEVELOPMENT.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+pgn-tactics-generator is a python application dedicated to creating chess puzzles/tactics from a pgn file.
+See [readme](./README.md) file for more details.
+
+
+# Dependencies
+
+This script requires the [Requests](https://docs.python-requests.org/) and [python-chess](https://python-chess.readthedocs.io/)
+libraries to run, as well as a copy of *Stockfish*.
+Is recommended that you use Python 3 and pip3. 
+It should work with Python 2.7 and pip (probably you will need to install futures `pip install futures` )
```
Comment (vitogit) [inline DEVELOPMENT.md:12]: Not sure if we support (or should support) python 2.7 anymore. So probably is best to just remove those lines and do the documentation based on python3 and pip3. What do you think?

**Review by vitogit — APPROVED** [review]
@karol-brejna-i  I just left a couple of simple comments but is approved. I added you as a collaborator, so you can merge it when you want. Thank you for this awesome PR and all the improvements!

File: DEVELOPMENT.md (hunk 6 above)
Comment (karol-brejna-i) [reply DEVELOPMENT.md:12]: I've created #28

**Pull request merged**